File: mobile/lib/models/sos_request.dart
import 'package:isar/isar.dart';

part 'sos_request.g.dart';

@collection
class SOSRequest {
  Id id = Isar.autoIncrement;

  @Index(unique: true)
  late String uuid;

  late double latitude;
  late double longitude;
  late double? altitude;
  late double? batteryLevel;
  
  late String category; // MEDICAL, LOST, FIRE, GENERAL
  
  @Index()
  late DateTime createdAt;

  late String status; // pending, sent, received, dispatched

  String? responderName;
}

File: backend/app/models/sos_request.py
import uuid
from datetime import datetime
from sqlalchemy import String, Float, DateTime, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column
from backend.app.db.base import Base

class SOSRequest(Base):
    __tablename__ = "sos_requests"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    device_id: Mapped[str | None] = mapped_column(String(200), nullable=True)
    latitude: Mapped[float] = mapped_column(Float, nullable=False)
    longitude: Mapped[float] = mapped_column(Float, nullable=False)
    altitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    battery_level: Mapped[float | None] = mapped_column(Float, nullable=True)
    category: Mapped[str] = mapped_column(String(20), default="GENERAL") # MEDICAL, LOST, FIRE, GENERAL
    
    status: Mapped[str] = mapped_column(
        String(20), default="pending", nullable=False
    )  # pending (sent) / received (authority ack) / dispatched (help on way)
    
    responder_name: Mapped[str | None] = mapped_column(String(100), nullable=True)
    
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
